elastic/beats
# Refactor of metricbeat process-gathering metrics and system/process

## What does this PR do?

This is a total refactor of the code used to gather process metrics on metricbeat. This was done with a few goals:

- Depend less on upstream libraries such as `gosigar`
- Performance improvements
- Reduce use of the `MapStr` type
- Get rid of any danging `null zero` metrics by using the `Opt` library.
- Prepare for upcoming changes to metricset/input design by decoupling metrics-gathering code from the logic of metricbeat
- Make `hostfs` handling consistent, no more global hostfs setting

As of now, I haven't added any reviewers to this, as I'm still not done testing, and I imagine it's gonna take me a good day or two to just make CI happy. However, it's mostly done, and I'd at least like to have it up so CI can fail.

## How the new code works

In order to improve performance, we try to make as few iterations over a list of processes as possible. In order to accomplish this, the code frequently between generic OS-agnostic functions and OS-specific implementations. When `system/process` fetches a list of processes, the steps are something like this:

- A call to `Get()` from an external library, such as metricbeat
  - A call to an OS-Specific `FetchPids()` method. This fetches and iterates over the PIDs using whatever OS-specific methods are available, then for each pid:
    - Call `pidFill()`, an OS-Agnostic function that:
      - Calls an OS-Specific `GetInfoForPid` function that fills out basic process info (name and PID at minimum)
      - Decides if a process has been filtered out by user regex
      - Calls an OS-specific `FillPidMetrics()` function that fetches the rest of the process metrics
      - Calculates percentages and cgroups metrics if applicable
    - Append the completed PID to a hashmap (for internal tracking) and an array (for reporting)
  - Sorts and reduces the returned list of pids based on a user-supplied `top` setting
  - Transforms the remaining metrics to a list of `MapStr` types, and returns 

This contains no user-facing changes.

## Why is it important?

This closes https://github.com/elastic/beats/issues/29796 and represents a major part of the ongoing cleanup of system code and further deprecation of of external libraries.

## Checklist



## Author's Checklist

- [ ] Manually run + check integration tests in linux, MacOS and Windows

## How to test this PR locally

- Pull down PR
- Build
- Run metricbeat with system/process, change various flags, make sure everything works and the data looks correct.
